### ISSUE TITLE
On openwrt only works when the iface is in promisc mode

### ISSUE THREAD

**Issue by ldrolez** [DanielAdolfsson/ndppd#13]
When running on my openwrt router, ndppd does not seem to see the "neighbor solicitation" from the external interface "br-wan".

But when I start a "tcpdump -I br-wan ...", the solicitations are now seen by the daemon and everything is proxyfied and I can ping the Internet.

Example solicitation sent:
19:42:54.451027 IP6 fe80::aaaa:bbbb:cccc:6cdf > ff02::1:ff00:1: ICMP6, neighbor solicitation, who has 2a01:xxxx...., length 32

Only when I start tcpdump, ndppd see them:

(debug) iface::read() len=86
(debug) iface::read_solicit() saddr=fe80::aaaa:bbbb:cccc:6cdf, daddr=ff02::1:ff00:1, len=86
(debug) proxy::handle_solicit() saddr=fe80::aaaa:bbbb:cccc:6cdf, taddr=2a01:xxxx....
(debug) checking 2a01:...::/64 against 2a01:xxxx...
(debug) session::create() pr=8468e0, saddr=fe80::aaaa:bbbb:cccc:6cdf, daddr=ff02::1:ff00:1, taddr=2a01:xxxx... =847368
(debug) iface::write_advert() daddr=fe80::aaaa:bbbb:cccc:6cdf, taddr=2a01:xxxx...
(debug) iface::write() daddr=fe80::aaaa:bbbb:cccc:6cdf, len=32

**Comment by nirgal** [comment]
I have the same problem on Debian.

I worked around it by forcing the promisc mode on boot. Here'a an extract of /etc/network/interfaces:
```
auto ethwan
iface ethwan inet dhcp
	up ifconfig ethwan promisc
```
Really ugly, right?

**Comment by DanielAdolfsson** [comment]
Strange indeed.

What kernels are you two running?

**Comment by ldrolez** [comment]
Linux OpenWrt 3.18.20 #1 Fri Sep 4 21:55:57 CEST 2015 mips GNU/Linux

**Comment by DanielAdolfsson** [comment]
Can you check if allmulti is enabled on the interface?
You can enable allmulti manually on the interface using:
`ifconfig ethwan allmulti`

**Comment by nirgal** [comment]
I'm using Debian stable: Kernel is "3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u4 (2016-02-29) x86_64 GNU/Linux"

`ifconfig ethwan allmulti`does fix the issue, just like `ifconfig ethwan promisc`does.

**Comment by DanielAdolfsson** [comment]
Interesting. ndppd is supposed to put the interface in allmulti mode, but for whatever reason it doesn't seem to do that.

**Comment by DanielAdolfsson** [comment]
The latest commit adds some additional logging to the allmulti() method. If it fails it should show up in the log now.

**Comment by nirgal** [comment]
I tried it and I don't have any extra logging:
```
Apr 18 09:25:11 phosphore ndppd[628]: (notice) ndppd (NDP Proxy Daemon) version 0.2.4
Apr 18 09:25:11 phosphore ndppd[628]: (notice) Using configuration file '/etc/ndppd.conf'
Apr 18 09:25:11 phosphore ndppd[628]: (warning) Low prefix length (120 <= 120) when using 'static' method
```

Here's my ndppd.conf file. Maybe I did something wrong:
```
proxy ethwan {
   rule 2a01:e34:ec10:4ef0::/120 {
	static
   }
}
```

**Comment by DanielAdolfsson** [comment]
You can start ndppd with **-vv** for debug verbosity.

For me it says:
`(debug) iface::allmulti() state=1, _name="ens3"`

If no errors are reported after this line, this should be reflected in ifconfig:
`UP BROADCAST RUNNING **ALLMULTI** MULTICAST  MTU:1500  Metric:1`

For clarity, ALLMULTI is not set after you start ndppd?
Also, you are running ndppd as root?

**Comment by DanielAdolfsson** [comment]
Also, when you tested allmulti, did you remove promisc?
You probably did, but I want to double-check :)

**Comment by nirgal** [comment]
Ouch, right, debug mode is not enabled by default. ^^
I am using your debian init.d script. I replaced `DAEMON_ARGS="-d -p $PIDFILE"` with `DAEMON_ARGS="-d -p $PIDFILE -vv"` and now I get:
```
Apr 18 12:37:51 phosphore ndppd[613]: (notice) ndppd (NDP Proxy Daemon) version 0.2.4
Apr 18 12:37:51 phosphore ndppd[613]: (notice) Using configuration file '/etc/ndppd.conf'
Apr 18 12:37:51 phosphore ndppd[613]: (debug) {
Apr 18 12:37:51 phosphore ndppd[613]: (debug)     proxy ethwan {
Apr 18 12:37:51 phosphore ndppd[613]: (debug)         rule 2a01:e34:ec10:4ef0::/120 {
Apr 18 12:37:51 phosphore ndppd[613]: (debug)             static
Apr 18 12:37:51 phosphore ndppd[613]: (debug)         }
Apr 18 12:37:51 phosphore ndppd[613]: (debug)     }
Apr 18 12:37:51 phosphore ndppd[613]: (debug) }
Apr 18 12:37:51 phosphore ndppd[613]: (debug) fd=3, hwaddr=0:e0:4c:69:9:14
Apr 18 12:37:51 phosphore ndppd[613]: (debug) iface::allmulti() state=1, _name="ethwan"
Apr 18 12:37:51 phosphore ndppd[634]: (debug) iface::~iface()
Apr 18 12:37:51 phosphore ndppd[634]: (debug) iface::allmulti() state=0, _name="ethwan"
Apr 18 12:37:51 phosphore ndppd[613]: (debug) proxy::create() if=ethwan
Apr 18 12:37:51 phosphore ndppd[613]: (warning) Low prefix length (120 <= 120) when using 'static' method
Apr 18 12:37:51 phosphore ndppd[613]: (debug) rule::create() if=ethwan, addr=2a01:e34:ec10:4ef0::/120, auto=no
Apr 18 12:37:51 phosphore ndppd[613]: ndppd[634]: (debug) iface::~iface()
Apr 18 12:37:51 phosphore ndppd[613]: ndppd[634]: (debug) iface::allmulti() state=0, _name="ethwan"
Apr 18 12:37:51 phosphore ndppd[639]: (debug) iface::fixup_pollfds() _map.size()=1
Apr 18 12:37:51 phosphore ndppd[639]: (debug) reading routes
(...)
```

ps shows that ndppd is running as root.

I removed any hack from /etc/network/interfaced for that interface, and I rebooted that server.

`ifconfig ethwan` shows:
 ```         UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1```
 and the proxy does NOT work.

If I enter:
```$ ifconfig ethwan multi```
then I get
``` UP BROADCAST RUNNING ALLMULTI MULTICAST  MTU:1500  Metric:1```
and the proxy works ok.

**Comment by DanielAdolfsson** [comment]
Ah I see the problem now.
`Apr 18 12:37:51 phosphore ndppd[634]: (debug) iface::~iface()`

ndppd is daemonizing, resulting in the iface destructor being called twice. I'll see if I can defer the interface setup until after daemonizing the process.

Thanks for the help!

**Comment by DanielAdolfsson** [comment]
Okay the latest commit defers the configuration until after ndppd is daemonized.
Please see if it works :)

**Comment by nirgal** [comment]
It works for me! :)
Thank you.

**Comment by DanielAdolfsson** [comment]
Cool.
Will do some additional testing and run it through valgrind. If it works, I'll bump the ndppd version.

Again, thanks for your help :)

**Comment by blbradley** [comment]
Hello!

Thanks for reporting this. ndppd v2.4.0 ships with Ubuntu 16.04 and is affected by this issue.

**Comment by nirgal** [comment]
@blbradley 

Indeed, since it was fixed in version 0.2.5. There is not much that can be done on the upstream part since having obsolete or unpatched version this is an Ubuntu specific problem.

Maybe you can try installing the Debian version 0.2.5-1 wich was released in april 2016, and might be compatible?
https://tracker.debian.org/media/packages/n/ndppd/changelog-0.2.5-3
Download the .deb file at http://snapshot.debian.org/package/ndppd/0.2.5-1/ and try installing it with dpkg -i. Don't try installing too many dependencies manually, if it doesn't work.

**Comment by blbradley** [comment]
I just enabled allmulti on that interface manually and in the config file for boot. Smooth sailing from here!
